utils/datasets/cifar10_dataset.py

```diff
@@ -0,0 +1,330 @@
+
+CIFAR_SHAPE = (32, 32, 3)
+TOTAL_FEATURE_SIZE = 32 * 32 * 3
+NUM_EXAMPLES_PER_CLIENT = 5000
+
+def load_cifar10_federated(
+    num_clients: int = 10, 
+    num_classes: int = 10, 
+    alpha: float = 1, 
+    train_client_batch_size: int = 20, 
+    test_client_batch_size: int = 100):
+  '''Construct a synthetic federated CIFAR-10 from its centralized version
+
+  Loads the train dataset into a non iid distribution over clients using the
+  sampling method based on LDA, taken from this paper:
+  Measuring the Effects of Non-Identical Data Distribution for
+  Federated Visual Classification (https://arxiv.org/pdf/1909.06335.pdf).
+
+  Args:
+    num_clients: An integer specifing the total number of clients.
+    num_classes: An integer representing the number of lables in the dataset.
+    alpha: A float controling the data heterogeneity among clients.
+    train_client_batch_size: A float representing the batch size during 
+      training.
+    test_client_batch_size: A float representing the batch size during
+      test.
+
+  Returns:
+    A tuple of `tff.simulation.ClientData` representing unpreprocessed
+    train data and test data.
+  '''
+  (train_images, train_labels), (test_images, test_labels) \
+      = datasets.cifar10.load_data()
+  client_train = collections.OrderedDict()
+  client_test = collections.OrderedDict()
+
+  # The following structure will store the image index for each client
+  idx_batch = [[] for _ in range(num_clients)]
+
+  for k in range(num_classes):
+    # For each class(label) sample in a proportion for each client. 
+    # The proportion is determined by a dirichlet distribution.
+    label_k = np.where(train_labels.squeeze()==k)[0]
+    np.random.shuffle(label_k)
+    proportion = np.random.dirichlet(alpha*np.ones((num_clients,)))
+    proportion = proportion*(label_k.shape[0])
+    proportion = np.cumsum(proportion).astype(int)
+    split_labels = np.split(label_k, proportion)
+      
+    idx_batch = [idx_j + splitk.tolist() for idx_j, splitk \
+       in zip(idx_batch, split_labels)]
+
+  num_per_client_test = int((len(test_labels) \
+      // (num_clients*test_client_batch_size))*test_client_batch_size)
+
+  for i in range(num_clients):
+    client_name = str(i)
+    x_train = train_images[np.array(idx_batch[i])]
+    y_train = train_labels[np.array(idx_batch[i])].astype('int64').squeeze()
+    train_samples_per_client = (x_train.shape[0] \
+        // train_client_batch_size) * train_client_batch_size
+    x_train = x_train[:train_samples_per_client]
+    y_train = y_train[:train_samples_per_client]
+    
+    data = collections.OrderedDict((('image', x_train), ('label', y_train)))
+    test_idx_start = i*num_per_client_test
+    test_idx_end = (i+1)*num_per_client_test       
+    test_data = collections.OrderedDict(
+        (('image', test_images[test_idx_start:test_idx_end]), 
+        ('label', test_labels[test_idx_start:test_idx_end].astype('int64').squeeze())))
```

Comment:
It seems that we're simply partitioning the test examples in order across clients. Couldn't we use LDA here too?